tests/e2e/main_test.go

```diff
@@ -312,3 +345,141 @@ func TestDefaultCollectorMetricsAvailable(t *testing.T) {
 		t.Fatalf("failed to find metrics of resources: %s", strings.Join(s, ", "))
 	}
 }
+
+func initClient(kubeconfigPath string) (*kubernetes.Clientset, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create config from kubeconfig, err=%v", err)
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client from config, err=%v", err)
+	}
+	return client, nil
+}
+
+func validateMetrics(metrics map[string]*dto.MetricFamily, nodeName string) error {
+	// A stable pod metric with node label
+	metricName := "kube_pod_info"
+	var metric *dto.MetricFamily
+	fmt.Println("## nodename", nodeName)
+
+	for _, m := range metrics {
+		if m.GetName() == metricName {
+			metric = m
+			break
+		}
+	}
+	if metric == nil {
+		return fmt.Errorf("metric %s isn't found", metricName)
+	}
+	for _, v := range metric.Metric {
+		for _, l := range v.GetLabel() {
+			if *l.Name == "node" && *l.Value != nodeName {
+				return fmt.Errorf("found new node name value %s", *l.Value)
+			}
+		}
+	}
+	return nil
+}
+
+func TestSharingPerNodeName(t *testing.T) {
+	if clientset == nil {
+		t.Skip("clientset is empty. Skip this test")
+	}
+
+	nodeName := ""
+	deploymentReplics := 1
+	expectedGeneration := 2
+	deploymentsClient := clientset.AppsV1().Deployments("kube-system")
+	nodesClient := clientset.CoreV1().Nodes()
+	podsClient := clientset.CoreV1().Pods("kube-system")
+
+	nodeList, listErr := nodesClient.List(context.TODO(), metav1.ListOptions{})
+	if listErr != nil {
+		t.Fatalf("failed to get latest version of Deployment: %v", listErr)
+	}
+	nodeNum := len(nodeList.Items)
+	fmt.Println("## nodeNum", nodeNum)
+
+	// 1. add --nodename and --resource to kube-state-metrics-agent
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		result, getErr := deploymentsClient.Get(context.TODO(), "kube-state-metrics", metav1.GetOptions{})
+		if getErr != nil {
+			return fmt.Errorf("failed to get kube-state-metrics deployment: %v", getErr)
+		}
+		if *result.Spec.Replicas != int32(deploymentReplics) {
+			return fmt.Errorf("Kube-state-metrics deployment replica != 1")
+		}
+		containers := result.Spec.Template.Spec.Containers
+		for idx, c := range containers {
+			if c.Name == "kube-state-metrics" {
+				containers[idx].Args = append(containers[idx].Args, "--nodename=$(NODE_NAME)")
+				containers[idx].Args = append(containers[idx].Args, "--resources=pods")
+				containers[idx].Env = append(containers[idx].Env, apiv1.EnvVar{
+					Name: "NODE_NAME",
+					ValueFrom: &apiv1.EnvVarSource{
+						FieldRef: &apiv1.ObjectFieldSelector{
+							APIVersion: "v1",
+							FieldPath:  "spec.nodeName",
+						},
+					},
+				})
+				break
+			}
+		}
+		_, updateErr := deploymentsClient.Update(context.TODO(), result, metav1.UpdateOptions{})
+		return updateErr
+	})
+	time.Sleep(time.Second * 100)
```

Comment:
What is the purpose of this sleep statement? Can we use something like `wait.Poll`?